Cropping site icon should preserve attachment properties

Rcreators opened the pull request:
<!--
Hi there! Thanks for contributing to WordPress!

Pull Requests in this GitHub repository **must** be linked to a ticket in the WordPress Core Trac instance (https://core.trac.wordpress.org), and are only used for code review. **No pull requests will be merged on GitHub.**

See the WordPress Handbook page on using PRs for Code Review more information: https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/

If this is your first time contributing, you may also find reviewing these guides first to be helpful:
- FAQs for New Contributors: https://make.wordpress.org/core/handbook/tutorials/faq-for-new-contributors/
- Contributing with Code Guide: https://make.wordpress.org/core/handbook/contribute/
- WordPress Coding Standards: https://make.wordpress.org/core/handbook/best-practices/coding-standards/
- Inline Documentation Standards: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/
- Browser Support Policies: https://make.wordpress.org/core/handbook/best-practices/browser-support/
- Proper spelling and grammar related best practices: https://make.wordpress.org/core/handbook/best-practices/spelling/
-->

As per the discussion over the ticket, Created the function wp_copy_parent_attachment_properties() in image.php file and called it in ajax-actions.php file.

Trac ticket: https://core.trac.wordpress.org/ticket/60524

---
**This Pull Request is for code review only. Please keep all other discussion in the Trac ticket. Do not merge this Pull Request. See [GitHub Pull Requests for Code Review](https://make.wordpress.org/core/handbook/contribute/git/github-pull-requests-for-code-review/) in the Core Handbook for more details.**

Rcreators committed:
Update class-wp-site-icon.php

Extended create_attachment_object() function to copy original attachment title, desc, alt text and caption to new cropped site icon file as well.

Rcreators committed:
Update class-wp-site-icon.php

Rcreators committed:
Ajax-actions.php and image.php file updated

Created a new copy attachment properties function in image.php file. Called it in ajax-actions.php file instead of long separate code for site-icon and default.

Rcreators committed:
Update ajax-actions.php

Rcreators committed:
Update image.php

Rcreators committed:
Update image.php

Removed extra white space

Rcreators committed:
Update image.php

Added missing variable

mukeshpanchal27 commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -4036,8 +4036,9 @@ function wp_ajax_crop_image() {
-			$attachment = $wp_site_icon->create_attachment_object( $cropped, $attachment_id );
-			unset( $attachment['ID'] );
+
+			/* Copy attachment properties */
```
```suggestion
			// Copy attachment properties.
```

For single line comment use these formate and add missing end period.

mukeshpanchal27 commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -4065,46 +4066,8 @@ function wp_ajax_crop_image() {
-					'_wp_attachment_image_alt' => wp_slash( $original_attachment->_wp_attachment_image_alt ),
-				);
-			}
+			/* Copy attachment properties */
```
```suggestion
			// Copy attachment properties.
```

mukeshpanchal27 commented on src/wp-admin/includes/image.php:
```diff
@@ -482,6 +482,60 @@ function _wp_make_subsizes( $new_sizes, $file, $image_meta, $attachment_id ) {
 }
 
+/**
+ * Copy parent attachment properties to newly cropped image.
+ *
+ *
+ * @param string $cropped Path to the cropped image file.
```
```suggestion
 * Copy parent attachment properties to newly cropped image.
 *
 * @since 6.5.0
 *
 * @param string $cropped Path to the cropped image file.
```

Add missing `@since`

Rcreators committed:
Apply suggestions from code review

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

Rcreators committed:
Cropping custom header image should preserve attachment properties

Same issue with custom image header cropping. Updated custom image header class to use newly created function to copy parent attachment properties.

joedolson committed:
Fix variable alignments

joedolson committed:
Eliminate unnecessary variable assignment.

joedolson committed:
Clarify variable comment slightly.

mukeshpanchal27 commented on src/wp-admin/includes/class-custom-image-header.php:
```diff
@@ -1421,7 +1423,11 @@ public function ajax_header_crop() {
 		$cropped = apply_filters( 'wp_create_file_in_uploads', $cropped, $attachment_id ); // For replication.
 
-		$attachment = $this->create_attachment_object( $cropped, $attachment_id );
+		$context = 'custom-header';
+
+		$attachment = wp_copy_parent_attachment_properties( $cropped, $attachment_id, $context );
```
```suggestion
		$attachment = wp_copy_parent_attachment_properties( $cropped, $attachment_id, 'custom-header' );
```

mukeshpanchal27 commented on src/wp-admin/includes/image.php:
```diff
@@ -482,6 +482,60 @@ function _wp_make_subsizes( $new_sizes, $file, $image_meta, $attachment_id ) {
+ *
+ * @since 6.5.0
+ *
+ * @param string $cropped Path to the cropped image file.
+ * @param int    $parent_attachment_id Parent file Attachment ID.
+ * @param string $context Control calling the function.
```
```suggestion
 * @param string $cropped              Path to the cropped image file.
 * @param int    $parent_attachment_id Parent file Attachment ID.
 * @param string $context              Control calling the function.
```

Alignments

Rcreators committed:
Apply suggestions from code review

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

joedolson commented on src/wp-admin/includes/ajax-actions.php:
```diff
@@ -4035,7 +4035,7 @@ function wp_ajax_crop_image() {
 			/** This filter is documented in wp-admin/includes/class-custom-image-header.php */
-			$cropped    = apply_filters( 'wp_create_file_in_uploads', $cropped, $attachment_id ); // For replication.
+			$cropped = apply_filters( 'wp_create_file_in_uploads', $cropped, $attachment_id ); // For replication.
 
```
Remove changes to spacing. I removed the spaces myself, but I hadn't noticed that the actual change was to add a line break between these lines of code, and it would be better to leave it unchanged.

joedolson commented on src/wp-admin/includes/class-custom-image-header.php:
```diff
@@ -1421,7 +1423,9 @@ public function ajax_header_crop() {
 
-		$attachment = $this->create_attachment_object( $cropped, $attachment_id );
+		$attachment = wp_copy_parent_attachment_properties( $cropped, $attachment_id, 'custom-header' );
+
```
Adjust to same alignment and spacing as elsewhere.

joedolson commented on src/wp-admin/includes/image.php:
```diff
@@ -482,6 +482,60 @@ function _wp_make_subsizes( $new_sizes, $file, $image_meta, $attachment_id ) {
+ *
+ * @param string $cropped              Path to the cropped image file.
+ * @param int    $parent_attachment_id Parent file Attachment ID.
+ * @param string $context              Control calling the function.
```
Let's make `$context` an optional argument with a default value of ''. While all the existing uses of this function require it, the function will have broader utility if it's optional.

joedolson reviewed (changes requested):
These are minor changes, but I think making the $context var optional would be valuable.

Rcreators committed:
Made context optional

mukeshpanchal27 commented on src/wp-admin/includes/image.php:
```diff
@@ -482,6 +482,60 @@ function _wp_make_subsizes( $new_sizes, $file, $image_meta, $attachment_id ) {
+	$url             = str_replace( wp_basename( $parent_url ), wp_basename( $cropped ), $parent_url );
+
+	$size       = wp_getimagesize( $cropped );
+	$image_type = ( $size ) ? $size['mime'] : 'image/jpeg';
```
```suggestion
	$image_type = $size ? $size['mime'] : 'image/jpeg';
```

Rcreators committed:
Update src/wp-admin/includes/image.php

Co-authored-by: Mukesh Panchal <mukeshpanchal27@users.noreply.github.com>

Rcreators committed:
Test Run

Rcreators committed:
Re Run

joedolson committed:
Deprecate create_attachment_object methods

joedolson committed:
Add unit test

joedolson committed:
Update wpCopyParentAttachmentProperties.php

joedolson committed:
Update wpCopyParentAttachmentProperties.php

joedolson committed:
Oy.

joedolson committed:
Update previous tests

joedolson committed:
Forgot I changed that

joedolson committed:
Pass context to icon & header; undefined var

joedolson committed:
Replace 2nd call to create_attachment_object

joedolson committed:
Set post parent in function directly instead of adding after

swissspidy commented on src/wp-admin/includes/class-custom-image-header.php:
```diff
@@ -1314,12 +1314,23 @@ final public function get_header_dimensions( $dimensions ) {
 	 * @return array An array with attachment object data.
 	 */
 	final public function create_attachment_object( $cropped, $parent_attachment_id ) {
+		_doing_it_wrong(
+			'custom_image_header_create_attachment_object',
+			sprintf(
+			// translators: 1: The Custom_Image_Header method that is no longer used by core. 2: The new function that replaces it.
+				__( 'The %1$s method has been replaced with %2$s.' ),
+				'Custom_Image_Header::create_attachment_object',
+				'wp_copy_parent_attachment_properties'
+			),
+			'6.5.0'
+		);
```
We have dedicated helpers for that :)

```suggestion
		_deprecated_function( __METHOD__, '6.5.0', 'wp_copy_parent_attachment_properties()' );
```

joedolson replied:
Huh. I did not know that existed. So much easier. Thanks!

swissspidy commented on src/wp-admin/includes/class-wp-site-icon.php:
```diff
@@ -78,12 +78,23 @@ public function __construct() {
 	 * @return array An array with attachment object data.
 	 */
 	public function create_attachment_object( $cropped, $parent_attachment_id ) {
+		_doing_it_wrong(
+			'wp_site_icon_create_attachment_object',
+			sprintf(
+			// translators: 1: The Custom_Site_Icon method that is no longer used by core. 2: The new function that replaces it.
+				__( 'The %1$s method has been replaced with %2$s.' ),
+				'WP_Site_Icon::create_attachment_object',
+				'wp_copy_parent_attachment_properties'
+			),
+			'6.5.0'
+		);
```
```suggestion
		_deprecated_function( __METHOD__, '6.5.0', 'wp_copy_parent_attachment_properties()' );

```

swissspidy committed:
Use `_deprecated_function()`

joedolson commented:
In r57755
